File: 3903-smallest-stable-index-i/3903-smallest-stable-index-i.cpp
class Solution {
public:
    int firstStableIndex(vector<int>& nums, int k) {
        int n=nums.size();
        vector<pair<int,int>>v(n);
        
        int mini=nums[n-1];
        int maxi=nums[0];
        for(int i=0;i<n;i++){
              if(nums[i]>=maxi){
                maxi=nums[i];
              }
              v[i].first=maxi;
             
        }
         for(int i=n-1;i>=0;i--){
              if(nums[i]<=mini){
                mini=nums[i];
              }
              v[i].second=mini;
             
        }


    for(int i=0;i<n;i++){
        if((v[i].first-v[i].second)<=k)return i;
    }

return -1;

    }
};

File: 3903-smallest-stable-index-i/3903-smallest-stable-index-i.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        mini=float('inf')
        maxi=float('-inf')
        n=len(nums)
        v=[[0,0] for _ in range(n)]
        for i in range(0,n,1):
           if nums[i]>maxi:
             maxi=nums[i]
           v[i][0]=maxi
        
        for i in range(n-1,-1,-1):
            if nums[i]<mini:
                mini=nums[i]
            v[i][1]=mini

        for i in range(0,n,1):
         if v[i][0]-v[i][1]<=k:return i

        return -1;

